tools/sof-ipc-timer.py

```diff
@@ -0,0 +1,149 @@
+    ppln_id: str
+    comp_id: int
+    wname: str
+
+    def __init__(self, ppln_id, comp_id, wname):
+        self.ppln_id = ppln_id
+        self.comp_id = comp_id
+        self.wname = wname
+
+    def __str__(self) -> str:
+        return f'{self.ppln_id}-{self.comp_id:#08x}'
+
+def read_comp_data(f):
+    '''Process the dmesg to get the component ID to component name mapping,
+    they are acquired from the line that contains 'Create widget':
+        [   59.622645] snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3:
+        Create widget host-copier.0.capture (pipe 3) - ID 4, instance 3, core 0"
+
+    By design in the kernel, pipeline ID is the instance ID of pipeline
+    widget, so it is acquired from the line that contains 'Create pipeline':
+        [   59.622134] snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3:
+        Create pipeline pipeline.3 (pipe 3) - instance 3, core 0
+
+    In practice, sof-test only capture kernel message and firmware trace generated during a test
+    case run. Mostly in manual tests, if the kernel message file contains multiple firmware runs
+    with overlapping information, the last one wins.
+    '''
+    comp_data = {}
+    ppln_id = None
+    for line in f:
+        if match_obj := re.search(r"Create (widget|pipeline)", line):
+            span_end_pos = match_obj.span()[1]
+            line_split = line[span_end_pos + 1:].split()
+            widget_name = line_split[0]
+            # In the linux kernel, IDA is used to allocated pipeline widget instance ID,
+            # this ID later is used for pipeline creation, thus becomes pipeline ID in the
+            # firmware. Note that ppln_id variable will be assigned properly at pipeline widget
+            # creation, because it is always the first one to be created before all other
+            # widgets in the same pipeline and pipelines are created sequentially.
+            if widget_name.startswith('pipeline'):
+                # remove ending comma(,) with [:-1]
+                ppln_id = int(line_split[5][:-1])
+                continue
+            # remove ending comma(,) with [:-1]
+            module_instance_id = int(line_split[7][:-1])
+            # remove ending comma(,) with [:-1]
+            widget_id = int(line_split[5][:-1])
+            # final module id are composed with high16(module instance id) + low16(module id)
+            widget_id |= module_instance_id << 16
+            comp_data[widget_id] = Component(ppln_id, widget_id, widget_name)
+    return comp_data
+
+def read_ipc_data(f, comp_data, args):
+    '''Pick ipc tx MOD_INIT_INSTANCE and MOD_LARGE_CONFIG_SET lines
+    from log, take time stamp from the first message and the 'done'
+    message, and store the difference of to either init_time or
+    conf_time in the component table.
+
+    '''
+    mod_id = None
+    start = None
+    for line in f:
+        if match_obj := re.search(r" ipc tx (     |reply|done )", line):
```

Comment:
The massive length of the indented block is a sure sign it should split into separate functions. Small functions also make interactive testing possible, that's what the `"__main__"` idiom below is for:

```shell
ipython3

>> import sof-ipc-timer
>> small_function("some log line")
```

Reply:
> The massive length of the indented block is a sure sign it should split into separate functions.

@lgirdwood 's comment just reminded me of something: naming is hard (c), but when you find good function names they act like a Table of Contents that lets you quickly get the gist of what is happening without reading the whole code/book.

Functions are not just about code re-use, they're useful even when you invoke them only once; for testability _and_ readability.

Reply:
Its hard to split the contents of the loop into separate functions, parsing of each line make changes to the variables variables to be used on later line. AFAIK there is no way to pass strings as reference in python... well I can try to rewrite the whole thing in object oriented manner,  but I am not sure if it becomes more readable that way.

Reply:
I think you could do at least something like this:
```python
    for line in f:
        if match_obj := re.search(r" ipc tx (     |reply|done )", line):
           (mod_id, and, other, global, state ) = big_line_parsing_function(mod_id, line, match_obj,...)
           # do something with (global, state) if needed
```
That should not require passing strings by reference or objects, should it?